In [1]:
!pip install -q \
langchain \
langchain-community \
langchain-google-genai \
langchain-text-splitters \
langchain-chroma \
chromadb \
wikipedia-api \
rank-bm25

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.8/43.8 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 33.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.4/69.4 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.3/23.3 MB 50.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.2/129.2 kB 10.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 18.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 89.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 48.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.7/18.7 MB 59.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.8/71.8 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.9/170.9 kB 13.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.3/61.

In [3]:
try:
    from google.colab import drive

    drive.mount('/content/drive')

    print("✅ Google Drive mounted successfully")

except Exception as e:
    print(f"❌ Drive Mount Error: {e}")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
✅ Google Drive mounted successfully


In [4]:
import os

try:
    from google.colab import userdata

    GEMINI_KEY = userdata.get("Gemini_key")

    if not GEMINI_KEY:
        raise ValueError(
            "Gemini_key not found in Colab Secrets"
        )

    os.environ["GOOGLE_API_KEY"] = GEMINI_KEY

    print(" Gemini API key loaded")

except Exception as e:
    print(f" API Key Error: {e}")

 Gemini API key loaded


Define project path

In [27]:
try:
    PROJECT_DIR = "/content/drive/MyDrive/RAG_Internship"

    WIKI_DIR = f"{PROJECT_DIR}/wiki_docs"

    CHROMA_DIR = "/content/phase4_chroma_db"
    print(" Paths configured")

    print("Wiki Folder:", WIKI_DIR)
    print("Chroma Folder:", CHROMA_DIR)

except Exception as e:
    print(f" Path Configuration Error: {e}")

 Paths configured
Wiki Folder: /content/drive/MyDrive/RAG_Internship/wiki_docs
Chroma Folder: /content/phase4_chroma_db


Load Documents

In [28]:
try:
    from langchain_community.document_loaders import (
        DirectoryLoader,
        TextLoader
    )

    loader = DirectoryLoader(
        WIKI_DIR,
        glob="*.txt",
        loader_cls=TextLoader
    )

    documents = loader.load()

    if len(documents) == 0:
        raise ValueError(
            "No documents found"
        )

    print(
        f" Loaded {len(documents)} documents"
    )

except Exception as e:
    print(f" Document Loading Error: {e}")

 Loaded 10 documents


In [29]:
try:
    print(
        documents[0].page_content[:1000]
    )
except Exception as e:
    print(f" Preview Error: {e}")

Natural language processing (NLP) is the processing of natural language information by a computer. NLP is a subfield of computer science and is closely associated with artificial intelligence. NLP is also related to information retrieval, knowledge representation, computational linguistics, and linguistics more broadly.
Major processing tasks in an NLP system include: speech recognition, text classification, natural language understanding, and natural language generation.

History
Natural language processing has its roots in the 1950s. Already in 1950, Alan Turing published an article titled "Computing Machinery and Intelligence," which proposed what is now called the Turing test as a criterion of intelligence, though at the time that was not articulated as a problem separate from artificial intelligence. The proposed test includes a task that involves the automated interpretation and generation of natural language.

Symbolic NLP (1950s – early 1990s)
The premise of symbolic NLP is oft

Better Chunking

In [30]:
try:

    from langchain_text_splitters import (
        RecursiveCharacterTextSplitter
    )

    splitter = RecursiveCharacterTextSplitter(
        chunk_size=1800,
        chunk_overlap=200,
        separators=[
            "\n\n",
            "\n",
            ". ",
            " ",
            ""
        ]
    )

    print(" Improved chunker created")

except Exception as e:

    print(f" Chunker Creation Error: {e}")

 Improved chunker created


In [31]:
try:

    chunks = splitter.split_documents(
        documents
    )

    print(
        f" Created {len(chunks)} chunks"
    )

except Exception as e:
    print(f" Chunking Error: {e}")

 Created 347 chunks


In [32]:
try:

    print(
        chunks[0].page_content
    )

except Exception as e:
    print(f" Chunk Preview Error: {e}")

Natural language processing (NLP) is the processing of natural language information by a computer. NLP is a subfield of computer science and is closely associated with artificial intelligence. NLP is also related to information retrieval, knowledge representation, computational linguistics, and linguistics more broadly.
Major processing tasks in an NLP system include: speech recognition, text classification, natural language understanding, and natural language generation.

History
Natural language processing has its roots in the 1950s. Already in 1950, Alan Turing published an article titled "Computing Machinery and Intelligence," which proposed what is now called the Turing test as a criterion of intelligence, though at the time that was not articulated as a problem separate from artificial intelligence. The proposed test includes a task that involves the automated interpretation and generation of natural language.

Symbolic NLP (1950s – early 1990s)
The premise of symbolic NLP is oft

Gemini Embeddings

In [33]:
try:

    from langchain_google_genai import (
        GoogleGenerativeAIEmbeddings
    )

    embeddings = (
        GoogleGenerativeAIEmbeddings(
            model="models/gemini-embedding-001"
        )
    )

    print(" Embedding model initialized")

except Exception as e:
    print(f" Embedding Initialization Error: {e}")

 Embedding model initialized


In [34]:
print(embeddings)

client=<google.genai.client.Client object at 0x7aef097e7590> model='models/gemini-embedding-001' task_type=None google_api_key=SecretStr('**********') credentials=None vertexai=None project=None location=None base_url=None additional_headers=None client_args=None api_version=None request_options=None output_dimensionality=None


Create ChoromaDB Database

In [35]:
try:

    import time
    import os
    import shutil

    from langchain_chroma import Chroma

    # Remove previous Phase4 DB
    if os.path.exists(CHROMA_DIR):

        shutil.rmtree(CHROMA_DIR)

        print(" Old ChromaDB removed")

    batch_size = 20

    total_chunks = len(chunks)

    print(f"Total chunks: {total_chunks}")

    db = None

    for start in range(
        0,
        total_chunks,
        batch_size
    ):

        end = min(
            start + batch_size,
            total_chunks
        )

        batch = chunks[start:end]

        print(
            f"\nProcessing batch "
            f"{start//batch_size + 1}"
            f" | chunks {start}-{end}"
        )

        try:

            if db is None:

                db = Chroma.from_documents(
                    documents=batch,
                    embedding=embeddings,
                    collection_name="wiki_rag_phase4",
                    persist_directory=CHROMA_DIR
                )

            else:

                db.add_documents(batch)

            print(
                f" Added {len(batch)} chunks"
            )

            # Wait to avoid rate limit
            time.sleep(15)

        except Exception as batch_error:

            print(
                f" Batch failed: {batch_error}"
            )

            print(
                "Waiting 65 seconds..."
            )

            time.sleep(65)

            db.add_documents(batch)

            print(
                " Retry successful"
            )

    print(
        "\n Phase 4 ChromaDB created successfully!"
    )

except Exception as e:

    print(
        f" ChromaDB Creation Error: {e}"
    )

Total chunks: 347

Processing batch 1 | chunks 0-20
 Added 20 chunks

Processing batch 2 | chunks 20-40
 Added 20 chunks

Processing batch 3 | chunks 40-60
 Added 20 chunks

Processing batch 4 | chunks 60-80
 Added 20 chunks

Processing batch 5 | chunks 80-100
 Added 20 chunks

Processing batch 6 | chunks 100-120
 Added 20 chunks

Processing batch 7 | chunks 120-140
 Added 20 chunks

Processing batch 8 | chunks 140-160
 Added 20 chunks

Processing batch 9 | chunks 160-180
 Added 20 chunks

Processing batch 10 | chunks 180-200
 Added 20 chunks

Processing batch 11 | chunks 200-220
 Added 20 chunks

Processing batch 12 | chunks 220-240
 Added 20 chunks

Processing batch 13 | chunks 240-260
 Added 20 chunks

Processing batch 14 | chunks 260-280
 Added 20 chunks

Processing batch 15 | chunks 280-300
 Added 20 chunks

Processing batch 16 | chunks 300-320
 Added 20 chunks

Processing batch 17 | chunks 320-340
 Added 20 chunks

Processing batch 18 | chunks 340-347
 Added 7 chunks

 Phase 4 Ch

In [36]:
try:

    count = db._collection.count()

    print(
        f" Collection Count: {count}"
    )

except Exception as e:

    print(f"{e}")

 Collection Count: 347


Save docs to Drive from colab temporary storage

In [37]:
import shutil
import os

DESTINATION = (
    "/content/drive/MyDrive/"
    "RAG_Internship/phase4_chroma_db"
)

try:

    if os.path.exists(DESTINATION):

        shutil.rmtree(DESTINATION)

    shutil.copytree(
        CHROMA_DIR,
        DESTINATION
    )

    print(
        " Phase4 ChromaDB saved to Drive"
    )

except Exception as e:

    print(
        f" Save Error: {e}"
    )

 Phase4 ChromaDB saved to Drive


Build BM25 Index

In [53]:
try:

    from rank_bm25 import BM25Okapi

    chunk_texts = [
        chunk.page_content
        for chunk in chunks
    ]

    tokenized_corpus = [
        text.lower().split()
        for text in chunk_texts
    ]

    bm25 = BM25Okapi(
        tokenized_corpus
    )

    print(
        f" BM25 built on {len(chunk_texts)} chunks"
    )

except Exception as e:

    print(
        f" BM25 Error: {e}"
    )

 BM25 built on 347 chunks


Vector Search

In [54]:
def vector_search(
    query,
    k=5
):

    try:

        docs = db.similarity_search(
            query,
            k=k
        )

        return docs

    except Exception as e:

        print(
            f"Vector Search Error: {e}"
        )

        return []

BM25 Search

In [55]:
import re

def bm25_search(
    query,
    k=5
):

    try:

        query_tokens = re.findall(
            r"\w+",
            query.lower()
        )

        scores = bm25.get_scores(
            query_tokens
        )

        ranked = sorted(
            zip(
                chunk_texts,
                scores
            ),
            key=lambda x: x[1],
            reverse=True
        )

        return ranked[:k]

    except Exception as e:

        print(
            f"BM25 Search Error: {e}"
        )

        return []

Hybrid Search

In [56]:
def hybrid_search(
    query,
    vector_k=3,
    bm25_k=2
):

    try:

        vector_docs = vector_search(
            query,
            vector_k
        )

        bm25_docs = bm25_search(
            query,
            bm25_k
        )

        final_docs = []

        seen = set()

        # Prioritize vector results
        for doc in vector_docs:

            text = doc.page_content

            if text not in seen:

                final_docs.append(doc)

                seen.add(text)

        # Add BM25 results
        for text, score in bm25_docs:

            if text not in seen:

                final_docs.append(text)

                seen.add(text)

        return final_docs

    except Exception as e:

        print(
            f"Hybrid Search Error: {e}"
        )

        return []

In [57]:
try:

    query = (
         "What are neural networks?"
    )

    results = hybrid_search(
        query
    )

    print(
        f"Retrieved {len(results)} chunks"
    )

except Exception as e:

    print(
        f"Test Error: {e}"
    )

Retrieved 5 chunks


In [60]:
try:

    for i, chunk in enumerate(
        results,
        start=1
    ):

        print("=" * 80)

        print(f"Chunk {i}")

        if hasattr(
            chunk,
            "page_content"
        ):

            print(
                chunk.page_content[:500]
            )

        else:

            print(
                chunk[:500]
            )

except Exception as e:

    print(
        f"Display Error: {e}"
    )

Chunk 1
A neural network is a group of interconnected units called neurons that send signals to one another. Neurons can be either biological cells or mathematical models. While individual neurons are simple, many of them together in a network can perform complex tasks. There are two main types of neural networks.

In neuroscience, a biological neural network is a physical structure found in brains and complex nervous systems – a population of nerve cells connected by synapses.
In machine learning, an a
Chunk 2
In machine learning
In machine learning, a neural network is an artificial mathematical model used to approximate nonlinear functions. While early artificial neural networks were physical machines, today they are almost always implemented in software.
Neurons in an artificial neural network are usually arranged into layers, with information passing from the first layer (the input layer) through one or more intermediate layers (the hidden layers) to the final layer (the output la

Retriever Function

In [67]:
!pip install -q groq

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.7/143.7 kB 3.5 MB/s eta 0:00:00


In [68]:
try:

    import os

    from groq import Groq

    from google.colab import userdata

    GROQ_API_KEY = userdata.get(
        "Groq_Key"
    )

    if not GROQ_API_KEY:

        raise ValueError(
            "Groq_Key not found in Colab Secrets"
        )

    client = Groq(
        api_key=GROQ_API_KEY
    )

    print(
        " Groq client initialized"
    )

except Exception as e:

    print(
        f" Groq Setup Error: {e}"
    )

 Groq client initialized


In [69]:
def ask_question(query):

    try:

        docs = hybrid_search(query)

        if len(docs) == 0:

            return (
                "No relevant documents found."
            )

        context_parts = []

        for doc in docs:

            if hasattr(
                doc,
                "page_content"
            ):

                context_parts.append(
                    doc.page_content
                )

            else:

                context_parts.append(
                    str(doc)
                )

        context = "\n\n".join(
            context_parts
        )

        prompt = f"""
You are a helpful assistant.

Answer ONLY using the provided context.

If the answer is not present
in the context, say:

'I could not find that information
in the documents.'

Context:
{context}

Question:
{query}
"""

        response = (
            client.chat.completions.create(
                model=
                "llama-3.3-70b-versatile",
                messages=[
                    {
                        "role":"user",
                        "content":prompt
                    }
                ]
            )
        )

        return (
            response
            .choices[0]
            .message
            .content
        )

    except Exception as e:

        return f"Error: {e}"

In [75]:
try:
    result = ask_question(
        "What is deep learning?"
    )

    print(result)

except Exception as e:

    print(f"❌ Error: {e}")

Deep learning refers to a class of machine learning algorithms in which a hierarchy of layers is used to transform input data into a progressively more abstract and composite representation. It utilizes multilayered neural networks to perform tasks such as classification, regression, and representation learning, and can learn which features to optimally place at which level on its own.


In [76]:
try:
    result = ask_question(
        "What is computer vision?"
    )

    print(result)

except Exception as e:

    print(f"❌ Error: {e}")

Computer vision is an interdisciplinary field that deals with how computers can be made to gain high-level understanding from digital images or videos. It seeks to automate tasks that the human visual system can do, and is concerned with the automatic extraction, analysis, and understanding of useful information from a single image or a sequence of images.


# Phase 4: RAG Enhancement

## Objective
The objective of Phase 4 is to improve the Retrieval-Augmented Generation (RAG) system by implementing retrieval enhancement techniques and evaluating their impact on answer quality.

---

## Implemented Improvements

### 1. Better Chunking (Technique D)
- Replaced the previous chunking strategy with an improved paragraph-aware and sentence-aware chunking approach.
- Used:
  - Chunk Size: 1800
  - Chunk Overlap: 200
- Generated 347 structured chunks from the AI knowledge base.

**Expected Benefits**
- Improved context preservation.
- Reduced information loss across chunk boundaries.
- Better retrieval quality.

### 2. Hybrid Search (Technique A)
Implemented a Hybrid Retrieval mechanism combining:

1. Vector Search (Semantic Retrieval)
   - Gemini Embeddings
   - ChromaDB Vector Store

2. BM25 Keyword Search
   - Rank-BM25 Retriever

The final retriever combines semantic similarity and keyword matching to retrieve more relevant context.

**Expected Benefits**
- Improved Context Recall.
- Improved Context Precision.
- Better handling of exact keyword queries.

---


## Technologies Used

- LangChain
- ChromaDB
- Gemini Embeddings
- Rank-BM25
- Groq Llama 3.3 70B
- Google Colab

---


The effectiveness of these improvements will be measured using RAGAS evaluation metrics in the Phase4_RAGAS_Evaluation notebook.